src/compiler/program.ts

```diff
@@ -5254,24 +5249,13 @@ export function getEmitModuleFormatOfFileWorker(sourceFile: Pick<SourceFile, "fi
 /** @internal Prefer `program.getImpliedNodeFormatForEmit` when possible. */
 export function getImpliedNodeFormatForEmitWorker(sourceFile: Pick<SourceFile, "fileName" | "impliedNodeFormat" | "packageJsonScope">, options: CompilerOptions): ResolutionMode {
     const moduleKind = getEmitModuleKind(options);
-    if (ModuleKind.Node16 <= moduleKind && moduleKind <= ModuleKind.NodeNext) {
-        return sourceFile.impliedNodeFormat;
-    }
     if (
-        sourceFile.impliedNodeFormat === ModuleKind.CommonJS
-        && (sourceFile.packageJsonScope?.contents.packageJsonContent.type === "commonjs"
-            || fileExtensionIsOneOf(sourceFile.fileName, [Extension.Cjs, Extension.Cts]))
+        ModuleKind.Node16 <= moduleKind && moduleKind <= ModuleKind.NodeNext
+        || fileExtensionIsOneOf(sourceFile.fileName, [Extension.Cts, Extension.Dcts, Extension.Cjs, Extension.Mts, Extension.Dmts, Extension.Mjs])
+        || pathContainsNodeModules(sourceFile.fileName)

```

Comment:
wouldn't this be problem for files that are symlinks from node_modules to path without node_modules?